### ISSUE TITLE
[REVIEW] llm-top-10: add self-learning feedback promotion and eval contamination gates

### ISSUE THREAD

**Issue by eldwin-easynet-world** [UnitOneAI/SecuritySkills#2384]
## Skill Being Reviewed
**Skill name:** llm-top-10
**Skill path:** `skills/ai-security/llm-top-10/`

## False Positive Analysis
**Benign code that triggers a false positive:**
```python
class FeedbackReviewQueue:
    def submit_feedback(self, user_id, case_id, label, evidence):
        # Feedback is evidence for a reviewer; it does not directly update
        # retrieval, prompts, routes, fine-tuning data, or model behavior.
        return review_queue.create(
            source_user=user_id,
            case_id=case_id,
            label=label,
            evidence_uri=evidence.uri,
            status="pending_human_review",
            can_affect_production=False,
        )
```

**Why this is a false positive:**
The current LLM04 guidance says to search for feedback loops that directly influence model behavior without review. That is right, but reviewers also need a way to distinguish safe feedback collection from unsafe self-learning. A queue that is pending review, cannot affect production behavior, and has no path into prompts, RAG, fine-tuning, routers, or memory should not be reported as a poisoning issue just because it stores user feedback.

## Coverage Gaps
**Missed variant 1: success/failure feedback directly rewrites production prompt policy**
```python
async def after_agent_run(run):
    lesson = await llm.extract_lesson(run.transcript, run.outcome)
    if run.outcome in ("success", "user_upvoted"):
        await prompt_policy_store.upsert(
            key="support_agent_operating_rules",
            value=lesson,
            source_run_id=run.id,
            status="active",
        )
```
**Why it should be caught:**
This is not only training-data poisoning or RAG document poisoning. The system is using outcome feedback as a production control-plane update. A manipulated transcript, brigaded thumbs-up signal, or incorrectly labeled "success" can become future operating policy. The current skill mentions RLHF/feedback loops, but it does not require reviewers to trace whether feedback can modify prompts, routing policies, system instructions, durable memory, tool-selection rules, or generated workflow/skill registries.

**Missed variant 2: evaluation failures are automatically promoted into training or memory, contaminating future safety metrics**
```python
failed_cases = eval_store.query(suite="security_jailbreaks", result="failed")

for case in failed_cases:
    # Intended as "learning from mistakes", but these cases remain in the
    # reported evaluation suite and are also promoted into supervised data.
    fine_tune_queue.append({
        "messages": case.messages,
        "expected": case.corrected_answer,
        "source": "eval_failure",
    })
    memory.upsert(case.attack_pattern, case.corrected_answer, trusted=True)
```
**Why it should be caught:**
The system can appear to improve while leaking evaluation prompts into training data or trusted memory. This can inflate safety scores and hide regressions. A robust review should require evidence that evaluation suites, feedback-derived training data, memory writebacks, and release-gating metrics are separated or explicitly marked so contaminated cases are excluded from reported performance.

**Missed variant 3: online routing learns from shallow success metrics without abuse resistance**
```ts
router.on("conversation_closed", async (event) => {
  if (event.csat >= 4 || event.resolved === true) {
    await routingPolicy.promote({
      promptTemplateId: event.promptTemplateId,
      modelId: event.modelId,
      tenant: "global",
    });
  }
});
```
**Why it should be caught:**
CSAT, closure, thumbs-up, conversion, or cost metrics can be manipulated or can optimize the wrong behavior. The review should check whether online learning and model/prompt routing use holdouts, canaries, abuse detection, rollback windows, and risk-specific guardrail metrics before promotion to global policy.

## Edge Cases
- A/B tests and bandit routers are legitimate, but they need guardrail metrics for safety, privacy, policy violations, and abuse, not only click-through, CSAT, conversion, or cost.
- User feedback should be scoped by tenant/user/product area; one customer segment's feedback should not silently become global behavior for regulated or high-risk tenants.
- "Learning from failures" is useful, but failed eval cases need lineage so they are not both training data and release-gating test data.
- Feedback labels from support agents, contractors, or downstream automation need identity, authority, and conflict handling; they should not be treated as equally trusted signals.
- Rollback must undo derived artifacts, not only the raw feedback row: prompt versions, embeddings, routing weights, memories, summaries, cached policies, and generated skills/workflows may all need invalidation.

## Remediation Quality
- [x] Fix resolves the vulnerability
- [x] Fix doesn't introduce new security issues
- [x] Fix doesn't break functionality
- **Issues found:** The current remediation says to add human review workflows and version-control knowledge bases, but it stops short of evidence gates for online self-learning systems that update prompts, memory, model routing, fine-tuning queues, or skill/workflow registries.

Recommended gates:
1. **Feedback-to-behavior lineage.** Require reviewers to map every feedback signal to the production artifacts it can change: prompt templates, system policies, RAG indexes, durable memory, model routers, fine-tuning data, adapters, generated skills, or workflows.
2. **Promotion controls.** Require approval state, source identity, sample size, abuse/brigading checks, holdout evaluation, safety guardrail metrics, and canary rollout before feedback-derived updates become global behavior.
3. **Train/eval separation.** Require release reports to mark feedback-derived cases and exclude contaminated evaluation examples from claimed safety/security metrics.
4. **Rollback and invalidation.** Require a tested rollback path for derived artifacts, including prompt versions, routing weights, memory entries, embeddings, summaries, caches, and generated procedures.

## Comparison to Other Tools
| Tool | Catches this? | Notes |
|------|:---:|-------|
| Semgrep | Partial | Can flag obvious direct writes from feedback handlers to prompt stores or training queues if project-specific patterns exist, but it will miss promotion semantics and evaluation contamination. |
| CodeQL | Partial | Can model taint from feedback inputs into policy/training stores, but custom queries are needed and won't capture canary, review, or guardrail requirements by default. |
| MLflow / W&B lineage | Partial | Can expose dataset/run lineage when teams record immutable artifacts, but it cannot determine whether promotion governance is safe. |
| This skill | Partial | It mentions RLHF/feedback loops, human review, and knowledge-base rollback, but does not provide concrete evidence gates for production self-learning and online policy updates. |

## Overall Assessment
**Strengths:**
- The skill correctly covers OWASP LLM04 data/model poisoning and calls out RLHF or feedback loops as a risk.
- It already pushes reviewers toward data provenance, human review, and knowledge-base rollback.
- The common pitfalls section is strong on end-to-end data flow awareness.

**Needs improvement:**
- Feedback loops are described too generally; the skill should tell reviewers how to trace feedback into prompts, routers, durable memory, training queues, generated skills, and workflows.
- It does not distinguish safe feedback collection from unsafe autonomous promotion.
- It lacks release-gating checks for train/eval contamination caused by "learn from failed evals" loops.

**Priority recommendations:**
1. Add a subsection under LLM04 for **self-learning feedback and online promotion controls**.
2. Add detection patterns for `feedback`, `thumbs_up`, `csat`, `conversation_closed`, `promote`, `routingPolicy`, `prompt_policy`, `fine_tune_queue`, `eval_failure`, `memory.upsert`, `trusted=True`, and generated `skill`/`workflow` registries.
3. Add an evidence table with `Signal`, `Source authority`, `Affected artifact`, `Promotion gate`, `Holdout/canary`, `Safety guardrail`, `Train/eval separation`, and `Rollback artifacts`.

## Bounty Info
- [x] I have read and agree to the [CONTRIBUTING.md](../../CONTRIBUTING.md) bounty terms
- **Preferred payment method:** GitHub Sponsors or PayPal; payment details can be provided privately after maintainer acceptance.

**Comment by Errordog2** [comment]
/attempt #2384

Submitted PR: https://github.com/UnitOneAI/SecuritySkills/pull/2385

Scope: adds LLM04 self-learning feedback promotion gates, benign feedback-queue false-positive guidance, train/eval contamination checks, reviewer detection markers, and rollback evidence requirements for the llm-top-10 skill.

Validation completed:
- git diff --check
- Markdown fenced-code balance check
- Targeted marker check for the new promotion, lineage, train/eval, memory, routing, and canary guidance

Requested bounty tier: moderate skill improvement (). Payment details can be provided privately after acceptance.
